### PR TITLE
Faster readlines

### PR DESCRIPTION
**Pull request by nirvdrum**
The basic idea here is to modify `readlines` so it doesn't call `gets`, which updates `$_` (wrong here), and doesn't need to `break` out of its iteration, which is quite expensive.

### CODE REVIEW

**Comment by nirvdrum** [comment]
These changes shave ~25% off the benchmark time in #792.

File: src/main/ruby/core/io.rb [hunk 1]
```diff
@@ -562,13 +562,14 @@ def self.foreach(name, separator=undefined, limit=undefined, options=undefined)
       io = File.open(name, options)
     end
 
+    each_reader = Truffle.privately { io.create_each_reader(separator, limit) }
+
     begin
-      while line = io.gets(separator, limit)
-        yield line
-      end
+      each_reader&.each { |line| yield line }
```
Comment (eregon) [inline src/main/ruby/core/io.rb]: Could we just pass the block argument (say `&block` here) so there is no extra block in between?

File: src/main/java/org/truffleruby/core/support/ByteArrayNodes.java [hunk 2]
```diff
@@ -156,19 +158,60 @@ public int setLength(DynamicObject bytes, int length) {
...
-            if (index == -1) {
+            if (notFoundProfile.profile(index == -1)) {
                 return nil();
             } else {
                 return index + patternRope.characterLength();
             }
         }
 
+        protected boolean isSingleBytePattern(DynamicObject pattern) {
+            assert RubyGuards.isRubyString(pattern);
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/support/ByteArrayNodes.java]: This assert is already done by `StringOperations.rope` FWIW.

File: src/main/ruby/core/io.rb [hunk 3]
```diff
@@ -272,12 +272,12 @@ def unseek!(io)
...
     # Returns +count+ bytes from the +start+ of the buffer as a new String.
     # If +count+ is +nil+, returns all available bytes in the buffer.
-    def shift(count=nil)
+    def shift(count=nil, encoding=Encoding::ASCII_8BIT)
       Truffle.synchronize(self) do
         total = size
         total = count if count and count < total
 
-        str = String.from_bytearray @storage, @start, total
+        str = String.from_bytearray @storage, @start, total, encoding
```
Comment (eregon) [inline src/main/ruby/core/io.rb]: Is there a use-case for a non-binary encoding?
I guess `shift` here is only called with 1 argument, but maybe another optimization can take advantage of it?

**Comment by chrisseaton** [comment]
Sorry should have waited for others' reviews.

File: src/main/ruby/core/io.rb (hunk 3 above)
Comment (nirvdrum) [reply src/main/ruby/core/io.rb]: I have a second set of changes where I create a UTF-8 string directly from the byte array. But it was starting to be a lot of changes for a single pull request, so I decided to split them.

File: src/main/ruby/core/io.rb (hunk 1 above)
Comment (nirvdrum) [reply src/main/ruby/core/io.rb]: That ought be doable. I'll look at that.

**Pull request merged**